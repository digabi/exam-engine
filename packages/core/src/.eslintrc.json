{
  "rules": {
    "no-restricted-imports": [
      "error",
      {
        "paths": [
          {
            "name": "lodash",
            "message": "Please use import * as _ from 'lodash-es' in the core module for better tree shaking."
          },
          {
            "name": "react-i18next",
            "importNames": ["useTranslation", "Translation"],
            "message": "Please use the `useExamTranslation` wrapper instead."
          }
        ]
      }
    ]
  }
}
