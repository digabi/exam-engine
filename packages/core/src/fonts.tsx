import FontFaceObserver from 'fontfaceobserver'
import * as _ from 'lodash-es'

/** A promise that resolves when the main fonts on the exam have been loaded. */
export const ready = Promise.all([
  new FontFaceObserver('Noto Sans', { weight: 300 }).load(), // Light
  new FontFaceObserver('Noto Sans', { weight: 400 }).load(), // Regular
  new FontFaceObserver('Noto Sans', { weight: 600 }).load(), // Semibold
])
  // eslint-disable-next-line promise/always-return
  .then(() => {
    loaded = true
  })
  .catch(_.noop)

/** A boolean flag that tells if the main fonts have been loaded.  */
export let loaded = false
