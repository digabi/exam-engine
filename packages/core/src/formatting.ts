export function formatQuestionDisplayNumber(displayNumber: string): string {
  return displayNumber.includes('.') ? displayNumber : displayNumber + '.'
}
