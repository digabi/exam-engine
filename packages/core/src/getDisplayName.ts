import React from 'react'

/** Returns the display name of a React component. Useful when creating HOCs. */
export function getDisplayName(WrappedComponent: React.ComponentType<any>): string {
  return WrappedComponent.displayName || WrappedComponent.name || 'Component'
}
