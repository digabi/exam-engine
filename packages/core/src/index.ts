import '../src/css/main.less'
import Attachments from './components/attachments/Attachments'
import Exam from './components/exam/Exam'
import GradingInstructions from './components/grading-instructions/GradingInstructions'
import Results from './components/results/Results'
import parseExam from './parser/parseExam'

export { Attachments, Exam, GradingInstructions, Results, parseExam }
export { ExamComponentProps } from './createRenderChildNodes'
export * from './types/ExamAnswer'
export * from './types/ExamServerAPI'
export * from './types/GradingStructure'
export * from './types/Score'

export interface ExamBundle {
  Attachments: typeof Attachments
  Exam: typeof Exam
  parseExam: typeof parseExam
  Results: typeof Results
}
