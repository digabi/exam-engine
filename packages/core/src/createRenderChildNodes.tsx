import React from 'react'
import { mapChildNodes } from './dom-utils'

export const ExamNamespaceURI = 'http://ylioppilastutkinto.fi/exam.xsd'
export const XHTMLNamespaceURI = 'http://www.w3.org/1999/xhtml'

export interface ExamComponentProps {
  className?: string
  /** An element in the exam XML. */
  element: Element
  /** A function that knows how to render the child nodes of this element. */
  renderChildNodes: RenderChildNodes
}

export const enum RenderOptions {
  RenderHTML,
  SkipHTML,
}

export type RenderChildNodes = (element: Element, options?: RenderOptions) => React.ReactNode[]

export function createRenderChildNodes(
  componentMap: Record<string, React.ComponentType<ExamComponentProps>>
): RenderChildNodes {
  function renderChildNode(node: ChildNode, index: number, options: RenderOptions): React.ReactNode {
    switch (node.nodeType) {
      case Node.TEXT_NODE:
      case Node.CDATA_SECTION_NODE:
        return options === RenderOptions.RenderHTML ? renderTextNode(node as Text) : null
      case Node.ELEMENT_NODE:
        return renderElement(node as Element, index, options)
      default:
        return null
    }
  }

  function renderElement(element: Element, index: number, options: RenderOptions) {
    switch (element.namespaceURI) {
      case XHTMLNamespaceURI:
        return renderXHTMLElement(element as HTMLElement, index, options)
      case ExamNamespaceURI:
        return renderExamElement(element, index)
      default:
        throw new Error(`Unrecognized element: ${element.localName}`)
    }
  }

  function renderXHTMLElement<T extends HTMLElement>(element: T, index: number, options: RenderOptions) {
    const children = element.hasChildNodes() ? renderChildNodes(element, options) : null

    if (options === RenderOptions.RenderHTML) {
      const Tag = element.localName
      const props = htmlAttributes2props(element, index)
      return <Tag {...props}>{children}</Tag>
    } else {
      return children
    }
  }

  function renderExamElement(element: Element, index: number) {
    const Component = componentMap[element.localName]
    const className = element.getAttribute('class') || undefined
    return Component ? <Component {...{ element, className, renderChildNodes, key: key(element, index) }} /> : null
  }

  function renderChildNodes(element: Element, options: RenderOptions = RenderOptions.RenderHTML): React.ReactNode[] {
    return mapChildNodes(element, (childElement, i) => renderChildNode(childElement, i, options))
  }

  return renderChildNodes
}

function renderTextNode(node: Text) {
  return node.textContent!
}

function htmlAttributes2props<T extends HTMLElement>(element: T, index: number): React.HTMLProps<T> {
  const props: React.HTMLProps<T> = { key: key(element, index) }
  for (const attribute of element.attributes) {
    const name = attribute.localName === 'class' ? 'className' : (attribute.localName as keyof React.HTMLProps<T>)
    props[name] = attribute.value
  }
  return props
}

function key(element: Element, index: number): string {
  return `${element.localName}-${index}`
}
