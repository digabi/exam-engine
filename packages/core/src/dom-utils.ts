import * as _ from 'lodash-es'
import { ExamNamespaceURI } from './createRenderChildNodes'
import { useContext, useEffect, useState } from 'react'
import { TOCContext } from './components/context/TOCContext'

export const NBSP = '\u00A0'

export function parentElements(element: Element, selector?: Selector): Element[] {
  const predicate = selector != null ? mkPredicate(selector) : undefined
  const result = []
  while (element.parentElement != null) {
    element = element.parentElement
    if (predicate == null || predicate(element)) {
      result.push(element)
    }
  }
  return result
}

type Selector = string | string[] | ((element: Element) => boolean)

function mkPredicate(selector: Selector): (element: Element) => boolean {
  return typeof selector === 'function'
    ? (element: Element) => element.namespaceURI === ExamNamespaceURI && selector(element)
    : Array.isArray(selector)
      ? (element: Element) => element.namespaceURI === ExamNamespaceURI && selector.includes(element.localName)
      : (element: Element) => element.namespaceURI === ExamNamespaceURI && element.localName === selector
}

export function query(root: Element, selector: Selector): Element | undefined {
  const predicate = mkPredicate(selector)

  function go(element: Element): Element | undefined {
    for (let i = 0, length = element.children.length; i < length; i++) {
      const childElement = element.children[i]
      if (predicate(childElement)) {
        return childElement
      }
      const maybeElement = go(childElement)
      if (maybeElement != null) {
        return maybeElement
      }
    }
  }

  return go(root)
}

export function queryAll(root: Element, selector: Selector, recurse = true): Element[] {
  const predicate = mkPredicate(selector)
  const results: Element[] = []

  function go(element: Element) {
    for (let i = 0, length = element.children.length; i < length; i++) {
      const childElement = element.children[i]
      if (predicate(childElement)) {
        results.push(childElement)
        if (recurse) {
          go(childElement)
        }
      } else {
        go(childElement)
      }
    }
  }

  go(root)
  return results
}

export function queryAncestors(element: Element, selector: Selector): Element | undefined {
  const predicate = mkPredicate(selector)
  while (element.parentElement != null) {
    element = element.parentElement
    if (predicate(element)) {
      return element
    }
  }
}

export function findChildElement(element: Element, selector: Selector): Element | undefined {
  const predicate = mkPredicate(selector)

  for (let i = 0, length = element.children.length; i < length; i++) {
    const childElement = element.children[i]
    if (predicate(childElement)) {
      return childElement
    }
  }
}
export function getAttribute(element: Element, attributeName: string): string | undefined {
  return element.getAttribute(attributeName) ?? undefined
}

export function getNumericAttribute(element: Element, attributeName: string): number | undefined {
  const maybeValue = element.getAttribute(attributeName)
  return maybeValue != null ? Number(maybeValue) : undefined
}

export function getBooleanAttribute(element: Element, attributeName: string): boolean {
  return getAttribute(element, attributeName) === 'true'
}

export function mapChildNodes<T>(element: Element, fn: (childNode: ChildNode, index: number) => T): T[] {
  const length = element.childNodes.length
  const result = new Array<T>(length)
  for (let i = 0; i < length; i++) {
    result[i] = fn(element.childNodes[i], i)
  }
  return result
}

export function mapChildElements<T>(element: Element, fn: (childElement: Element, index: number) => T): T[] {
  const result = []
  for (let i = 0, length = element.children.length; i < length; i++) {
    const childElement = element.children[i]
    result.push(fn(childElement, i))
  }
  return result
}

export function findChildrenAnswers(element: Element): Element[] {
  return queryAll(element, ['choice-answer', 'dropdown-answer', 'text-answer', 'scored-text-answer'], true)
}

export function hasSiblingQuestions(answer: Element) {
  const parentQuestion = queryAncestors(answer, 'question')
  const siblingQuestions = parentQuestion?.querySelectorAll('[question-id]')
  return siblingQuestions && siblingQuestions?.length > 1
}

export function createElement<K extends keyof HTMLElementTagNameMap>(
  tagName: K,
  properties?: Record<string, string> | null,
  ...children: Array<string | ChildNode>
): HTMLElementTagNameMap[K] {
  const element = document.createElement(tagName)
  setProperties(element, properties)
  children.forEach(c => element.appendChild(typeof c === 'string' ? new Text(c) : c))
  return element
}

export function setProperties(element: HTMLElement, properties?: Record<string, string> | null): void {
  _.forEach(properties, (v, k) => {
    if (k === 'css') {
      element.style.cssText = v
    } else if (k.startsWith('data-')) {
      element.setAttribute(k, v)
    } else {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
      ;(element as any)[k] = v
    }
  })
}

export function useIsElementInViewport(elementType: 'question' | 'section', displayNumber: string) {
  const { visibleTOCElements } = useContext(TOCContext)
  const [isVisible, setIsVisible] = useState(false)

  const id = `${elementType}-${displayNumber}`

  useEffect(() => {
    if (visibleTOCElements?.includes(id) && !isVisible) {
      setIsVisible(true)
    } else if (!visibleTOCElements?.includes(id) && isVisible) {
      setIsVisible(false)
    }
  }, [visibleTOCElements])

  return isVisible
}

export function getElementPath(element: Element): string {
  const temp = !element.tagName && Array.from(element.parentNode?.childNodes || []).indexOf(element)
  const elementIndex = Array.from(element.parentElement?.children || []).indexOf(element)
  let path = `${element.nodeName}:${elementIndex > 0 ? elementIndex : temp}`
  let currentElement = element
  while (currentElement.parentElement) {
    currentElement = currentElement.parentElement
    const displayNumber = currentElement.getAttribute('display-number')
    const index =
      Number(displayNumber) ||
      Array.from(currentElement.parentElement?.children || [currentElement]).indexOf(currentElement)
    const elementId = currentElement.getAttribute('id')
    path = `${elementId || currentElement.nodeName}:${index} > ${path}`
  }

  return path
}
