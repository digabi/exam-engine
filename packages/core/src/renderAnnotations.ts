import { Annotation, ImageAnnotation, TextAnnotation } from './types/Score'
import { createElement } from './dom-utils'
import classNames from 'classnames'

export function renderAnnotations(
  element: HTMLElement,
  pregradingAnnotations: Annotation[],
  censoringAnnotations: Annotation[]
): void {
  const annotations = [...pregradingAnnotations, ...censoringAnnotations]
  const annotationsWithMessages = annotations.filter((a) => a.message)

  for (const annotation of annotations) {
    const type = pregradingAnnotations.includes(annotation) ? 'pregrading' : 'censoring'
    const index = annotationsWithMessages.indexOf(annotation) + 1 || undefined

    switch (annotation.type) {
      case 'line':
      case 'rect':
        renderImageAnnotation(element, annotation, type, index)
        break
      default:
        renderTextAnnotation(element, annotation, type, index)
    }
  }
}

function renderImageAnnotation(
  element: HTMLElement,
  annotation: ImageAnnotation,
  type: 'pregrading' | 'censoring',
  index?: number
): void {
  const { attachmentIndex } = annotation
  const image = element.querySelectorAll('img')[attachmentIndex]

  const wrapper = getWrapper()
  const shape = mkShape()
  wrapper.appendChild(shape)

  function getWrapper() {
    const parent = image.parentElement!

    if (parent instanceof HTMLSpanElement) {
      return parent
    }

    const wrapper = createElement('span', { className: 'e-annotation-wrapper' })
    parent.insertBefore(wrapper, image)
    wrapper.appendChild(image)

    return wrapper
  }

  function mkShape() {
    const mark = createElement('mark', {
      className: classNames('e-annotation e-annotation--shape', {
        'e-annotation--pregrading': type === 'pregrading',
        'e-annotation--censoring': type === 'censoring',
      }),
      title: annotation.message,
    })

    const style = mark.style
    const pct = (n: number) => `${n * 100}%`

    if (annotation.type === 'rect') {
      style.left = pct(annotation.x)
      style.top = pct(annotation.y)
      style.right = pct(1 - (annotation.x + annotation.width))
      style.bottom = pct(1 - (annotation.y + annotation.height))
    } else {
      style.left = pct(annotation.x1)
      style.top = pct(annotation.y1)
      style.right = pct(1 - annotation.x2)
      style.bottom = pct(1 - annotation.y2)
    }

    if (index) {
      mark.appendChild(createSup(index, 'shape'))
    }

    return mark
  }
}

function renderTextAnnotation(
  rootElement: HTMLElement,
  annotation: TextAnnotation,
  type: 'pregrading' | 'censoring',
  index?: number
): void {
  const { startIndex, length: annotationLength } = annotation
  return go(rootElement.childNodes[0], 0, annotationLength)

  function go(
    /* The current node in the answer HTML. */
    node: ChildNode,
    /** The current index corresponding to the node. */
    currentIndex: number,
    /* The remaining number of characters from the annotation to render. */
    remaining: number
  ): void {
    if (!isMark(node) && !isSup(node) && !isBr(node) && currentIndex >= startIndex) {
      // We found the correct spot to start an annotation. So let's create a mark element.
      const mark = mkMark()
      // Insert it before the node.
      node.parentElement?.insertBefore(mark, node)
      // Move the node and its next siblings inside the mark until the mark
      // contains the required amount of content or we run out of sibling nodes.
      // If the node is a Text node and it is too long for the annotation, we
      // split it in two and move the first part inside the mark, leaving the
      // rest in place.
      const stillRemaining = move(mark, node, remaining)

      // If we have run out of sibling nodes but still have characters to mark,
      // create another mark at the next text node. This means that marks have
      // been nested.
      if (stillRemaining > 0) {
        go(nextSibling(mark), currentIndex + annotationLength - stillRemaining, stillRemaining)
      }
      // We know that we're at the last mark of this annotation.
      // Render the superscript index after it, if necessary.
      else if (index) {
        mark.appendChild(createSup(index, 'text'))
      }
    } else {
      if (node instanceof Text && currentIndex + length(node) > startIndex) {
        // The annotation should start in the middle of this text node. We need to split it in two.
        node.splitText(startIndex - currentIndex)
      }

      go(next(node), currentIndex + length(node), remaining)
    }
  }

  function isBr(node: ChildNode): node is HTMLBRElement {
    return node.nodeName === 'BR'
  }

  function isMark(node: ChildNode): node is HTMLElement {
    return node.nodeName === 'MARK'
  }

  function isSup(node: ChildNode): node is HTMLElement {
    return node.nodeName === 'SUP'
  }

  function isImage(node: ChildNode): node is HTMLImageElement | HTMLSpanElement {
    // If we've already added an image annotation to the image, it will be wrapped
    // in a <span> element.
    return node.nodeName === 'IMG' || node.nodeName === 'SPAN'
  }

  function next(node: ChildNode): ChildNode {
    return isMark(node) ? node.childNodes[0] : nextSibling(node)
  }

  function nextSibling(node: ChildNode): ChildNode {
    const sibling = node.nextSibling

    if (sibling) {
      return sibling
    }

    const parent = node.parentElement

    // If we encounter the root element, we've run out of text to annotate. This
    // could be a bug in the annotation code or perhaps the annotations are invalid.
    // To err on the side of caution, throw an exception in this case.
    if (parent === null || parent === rootElement) {
      throw new Error('Bug: the answer seems to be too short for the current set of annotations.')
    }

    return nextSibling(parent)
  }

  function length(node: ChildNode): number {
    return node instanceof Text ? node.length : isImage(node) ? 1 : 0
  }

  function mkMark(): HTMLElement {
    return createElement('mark', {
      className: classNames('e-annotation', {
        'e-annotation--pregrading': type === 'pregrading',
        'e-annotation--censoring': type === 'censoring',
      }),
      title: annotation.message,
    })
  }

  function move(mark: HTMLElement, node: ChildNode | null, remaining: number): number {
    if (!node || isMark(node) || remaining === 0) {
      return remaining
    }

    // Split text nodes if they are too long.
    if (node instanceof Text && length(node) > 0 && length(node) > remaining) {
      node.splitText(remaining)
    }

    const nextSibling = node.nextSibling
    mark.appendChild(node)
    return move(mark, nextSibling, remaining - length(node))
  }
}

function createSup(index: number, type: 'text' | 'shape') {
  const content = `${index})`
  const children = type === 'text' ? [] : content
  return createElement(
    'sup',
    {
      className: `e-annotation__index e-annotation__index--${type}`,
      'data-content': content,
    },
    ...children
  )
}
