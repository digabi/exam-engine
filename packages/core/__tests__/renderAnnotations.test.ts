import { renderAnnotations } from '../src/renderAnnotations'
import { Annotation } from '../src'
import wrap from 'jest-snapshot-serializer-raw'

describe('renderAnnotations()', () => {
  it('one text annotation', () => {
    expect(
      render(
        'Lorem ipsum dolor sit amet',
        [
          {
            startIndex: 0,
            length: 5,
            message: 'foo',
          },
        ],
        []
      )
    ).toMatchSnapshot()
  })

  it('two text annotations, side by side', () => {
    expect(
      render(
        'Lorem ipsum dolor sit amet',
        [
          { startIndex: 0, length: 5, message: 'foo' },
          { startIndex: 5, length: 5, message: 'bar' },
        ],
        []
      )
    ).toMatchSnapshot()
  })

  it('two text annotations with space in between', () => {
    expect(
      render(
        'Lorem ipsum dolor sit amet',
        [
          { startIndex: 0, length: 5, message: 'foo' },
          { startIndex: 10, length: 5, message: 'bar' },
        ],
        []
      )
    ).toMatchSnapshot()
  })

  it('overlapping pregrading and censoring annotations', () => {
    expect(
      render(
        'Lorem ipsum dolor sit amet',
        [{ startIndex: 5, length: 10, message: 'foo' }],
        [{ startIndex: 0, length: 10, message: 'bar' }]
      )
    ).toMatchSnapshot()
  })

  it('text annotation through an image', () => {
    expect(
      render(
        'Lorem <img src="" /> ipsum dolor sit amet',
        [
          {
            startIndex: 0,
            length: 10,
            message: 'foo',
          },
        ],
        []
      )
    ).toMatchSnapshot()
  })

  it('annotation starts after <br /> tags', () => {
    expect(
      render(
        '<br />Lorem ipsum dolor sit amet',
        [
          {
            startIndex: 0,
            length: 5,
            message: 'foo',
          },
        ],
        []
      )
    ).toMatchSnapshot()
  })

  it('rect annotation', () => {
    expect(
      render(
        `Lorem <img src="" /> ipsum dolor sit amet`,
        [
          {
            type: 'rect',
            attachmentIndex: 0,
            message: 'foo',
            x: 0.25,
            y: 0.25,
            height: 0.5,
            width: 0.5,
          },
        ],
        []
      )
    ).toMatchSnapshot()
  })

  it('line annotation', () => {
    expect(
      render(
        `Lorem <img src="" /> ipsum dolor sit amet`,
        [
          {
            type: 'line',
            attachmentIndex: 0,
            message: 'foo',
            x1: 0.25,
            y1: 0.25,
            x2: 0.75,
            y2: 0.25,
          },
        ],
        []
      )
    ).toMatchSnapshot()
  })

  it('checks the length of annotated images correctly in the text annotation code', () => {
    expect(
      render(
        `Lorem <img src="" /> ipsum dolor <img src="" /> sit amet`,
        [
          {
            type: 'rect',
            attachmentIndex: 0,
            message: 'First image annotation',
            x: 0,
            y: 0,
            width: 1,
            height: 1,
          },
          {
            type: 'rect',
            attachmentIndex: 1,
            message: 'First image annotation',
            x: 0,
            y: 0,
            width: 1,
            height: 1,
          },
          {
            type: 'text',
            startIndex: 25,
            message: 'Text annotation',
            length: 5,
          },
        ],
        []
      )
    ).toMatchSnapshot()
  })

  it('throws an error if the answer is too short for the annotations', () => {
    expect(() =>
      render(
        'Lorem ipsum dolor sit amet',
        [
          {
            startIndex: 5,
            length: 999,
            message: 'Boom!',
          },
        ],
        []
      )
    ).toThrow(new Error('Bug: the answer seems to be too short for the current set of annotations.'))
  })
})

function createAnswer(html: string) {
  const div = document.createElement('div')
  div.innerHTML = html
  return div
}

function render(
  html = 'Lorem ipsum dolor sit amet',
  pregradingAnnotations: Annotation[],
  censoringAnnotations: Annotation[]
) {
  const answer = createAnswer(html)
  renderAnnotations(answer, pregradingAnnotations, censoringAnnotations)
  return wrap(answer.innerHTML)
}
