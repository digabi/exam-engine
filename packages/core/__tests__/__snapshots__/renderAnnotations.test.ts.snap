// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderAnnotations() annotation starts after <br /> tags 1`] = `<br><mark class="e-annotation e-annotation--pregrading" title="foo">Lorem<sup class="e-annotation__index e-annotation__index--text" data-content="1)"></sup></mark> ipsum dolor sit amet`;

exports[`renderAnnotations() checks the length of annotated images correctly in the text annotation code 1`] = `Lorem <span class="e-annotation-wrapper"><img src=""><mark class="e-annotation e-annotation--shape e-annotation--pregrading" title="First image annotation" style="left: 0%; top: 0%; right: 0%; bottom: 0%;"><sup class="e-annotation__index e-annotation__index--shape" data-content="1)">1)</sup></mark></span> ipsum dolor <span class="e-annotation-wrapper"><img src=""><mark class="e-annotation e-annotation--shape e-annotation--pregrading" title="First image annotation" style="left: 0%; top: 0%; right: 0%; bottom: 0%;"><sup class="e-annotation__index e-annotation__index--shape" data-content="2)">2)</sup></mark></span> sit<mark class="e-annotation e-annotation--pregrading" title="Text annotation"> amet<sup class="e-annotation__index e-annotation__index--text" data-content="3)"></sup></mark>`;

exports[`renderAnnotations() line annotation 1`] = `Lorem <span class="e-annotation-wrapper"><img src=""><mark class="e-annotation e-annotation--shape e-annotation--pregrading" title="foo" style="left: 25%; top: 25%; right: 25%; bottom: 75%;"><sup class="e-annotation__index e-annotation__index--shape" data-content="1)">1)</sup></mark></span> ipsum dolor sit amet`;

exports[`renderAnnotations() one text annotation 1`] = `<mark class="e-annotation e-annotation--pregrading" title="foo">Lorem<sup class="e-annotation__index e-annotation__index--text" data-content="1)"></sup></mark> ipsum dolor sit amet`;

exports[`renderAnnotations() overlapping pregrading and censoring annotations 1`] = `<mark class="e-annotation e-annotation--censoring" title="bar">Lorem</mark><mark class="e-annotation e-annotation--pregrading" title="foo"><mark class="e-annotation e-annotation--censoring" title="bar"> ipsu<sup class="e-annotation__index e-annotation__index--text" data-content="2)"></sup></mark>m dol<sup class="e-annotation__index e-annotation__index--text" data-content="1)"></sup></mark>or sit amet`;

exports[`renderAnnotations() rect annotation 1`] = `Lorem <span class="e-annotation-wrapper"><img src=""><mark class="e-annotation e-annotation--shape e-annotation--pregrading" title="foo" style="left: 25%; top: 25%; right: 25%; bottom: 25%;"><sup class="e-annotation__index e-annotation__index--shape" data-content="1)">1)</sup></mark></span> ipsum dolor sit amet`;

exports[`renderAnnotations() text annotation through an image 1`] = `<mark class="e-annotation e-annotation--pregrading" title="foo">Lorem <img src=""> ip<sup class="e-annotation__index e-annotation__index--text" data-content="1)"></sup></mark>sum dolor sit amet`;

exports[`renderAnnotations() two text annotations with space in between 1`] = `<mark class="e-annotation e-annotation--pregrading" title="foo">Lorem<sup class="e-annotation__index e-annotation__index--text" data-content="1)"></sup></mark> ipsu<mark class="e-annotation e-annotation--pregrading" title="bar">m dol<sup class="e-annotation__index e-annotation__index--text" data-content="2)"></sup></mark>or sit amet`;

exports[`renderAnnotations() two text annotations, side by side 1`] = `<mark class="e-annotation e-annotation--pregrading" title="foo">Lorem<sup class="e-annotation__index e-annotation__index--text" data-content="1)"></sup></mark><mark class="e-annotation e-annotation--pregrading" title="bar"> ipsu<sup class="e-annotation__index e-annotation__index--text" data-content="2)"></sup></mark>m dolor sit amet`;
